### PR TITLE
adds support for W3C ES256K

### PR DESCRIPTION
**Pull request by scrummitch**
Adds support for decoding JWTs using EcdsaSecp256k1Signature2019 which will become very popular due to Decentralised Identifiers becoming a W3C recommendation.

These signatures are used in Verifiable Credential issuance and presentations, see here https://identity.foundation/jwt-vc-presentation-profile. 

Microsoft is using this method to sign their JWTs, see here https://learn.microsoft.com/en-us/azure/active-directory/verifiable-credentials/verifiable-credentials-standards#supported-algorithms 

Please let me know if I've missed anything.

### CODE REVIEW

**Comment by OR13** [comment]
I recommend changing "EcdsaSecp256k1VerificationKey2019" to ES256K. In the context of JWT, only ES256K is needed to signal the alg and key type.

**Comment by scrummitch** [comment]
> This looks great!
> 
> One question I have is where you got the value for the EC curve (`1.3.132.0.10`). I searched for it online, and only found it referenced in a few websites like oid-info.com and oidref.com. It would be nice to see this value referenced from somewhere like IETF or other standards organization.

@bshaffer I used oid-info.com but its also in the source code for phpseclib here https://github.com/phpseclib/phpseclib/blob/master/phpseclib/Crypt/EC/Formats/Keys/Common.php#L120

**Comment by scrummitch** [comment]
@bshaffer Would like to get your opinion on swapping the order of the merge in `JWT::encode()`. For example to so `typ` can be blanked out. RFC 7519 says its not required https://www.rfc-editor.org/rfc/rfc7519#section-5.1 

```php 
if (isset($head) && \is_array($head)) {
    $header = \array_merge($head, $header);
}
```

to 

```php 
if (isset($head) && \is_array($head)) {
    $header = array_filter(\array_merge($header, $head));
}
```

**Comment by scrummitch** [comment]
> > @bshaffer Would like to get your opinion on swapping the order of the merge in JWT::encode().
> 
> My initial thought is that this could change existing behavior, so unless it's a bug, I don't think we should allow it. Also, it's unrelated (AFAIK) to this PR, and so should be in a separate PR as a feat or fix.
> 
> What is the rational for blanking out `typ`?

Agreed it should be a separate PR so i'll raise that later when required

In the DIF .well-known DID configuration file the JWT used for proof requires `typ` to not exist in the header so i've been working backwards from that. https://identity.foundation/.well-known/resources/did-configuration/#json-web-token-proof-format 

Maybe @OR13 knows why this is a requirement ?

**Comment by OR13** [comment]
That requirement regarding typ is getting removed, I'd ignore it.

File: src/JWT.php [hunk 1]
```diff
@@ -132,8 +133,8 @@ public static function decode(
             // See issue #351
             throw new UnexpectedValueException('Incorrect key for this algorithm');
         }
-        if ($header->alg === 'ES256' || $header->alg === 'ES384') {
-            // OpenSSL expects an ASN.1 DER sequence for ES256/ES384 signatures
+        if (in_array($header->alg, ['ES256', 'ES256K', 'ES384'])) {
```
Comment (bshaffer) [inline src/JWT.php:136]: Fix style check and also make type strict
```suggestion
        if (\in_array($header->alg, ['ES256', 'ES256K', 'ES384'], true)) {
```

**Review by bshaffer — COMMENTED** [review]
Looking pretty good! Just one more suggestion and this is good to go.

EDIT: Also, it looks like you still need to sign the [Google CLA](https://cla.developers.google.com/)

**Comment by scrummitch** [comment]
@bshaffer thanks for the comment, ive updated that strict flag as per your suggestion.

You can see i've already accepted the CLA here: 
image

**Review by bshaffer — APPROVED** [review]

**Pull request merged**